### PR TITLE
Adds Sina Python Package to Spack

### PR DESCRIPTION
**Pull request by doutriaux1**

### CODE REVIEW

**Review by estebanpauli — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-sina/package.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-sina/package.py]:
```suggestion
# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```

File: var/spack/repos/builtin/packages/py-sina/package.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    without requiring knowledge of one, allowing queries to be expressed in pure
+    Python.  Visualizations are also provided through Python.
+
+    Sina is intended especially for use with run metadata,
+    allowing users to easily and efficiently find simulation runs that match some
+    criteria.
+    """
+
+    homepage = "https://github.com/LLNL/Sina"
+    git = "https://github.com/LLNL/Sina"
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-sina/package.py]:
```suggestion
    git = "https://github.com/LLNL/Sina.git"
```

File: var/spack/repos/builtin/packages/py-sina/package.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-ujson', type=('build', 'run'))
+    depends_on("py-sqlalchemy", type="run")
+    depends_on("py-six", type="run")
+
+    # Licensing
+    license_required = False
+    license_comment  = '#'
+    license_files    = ['LICENSE']
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-sina/package.py]: These are not needed unless the software requires a license (i.e., is commercial software).

```suggestion
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Are you able to build this package with `spack install py-sina`?  

I'm encountering problems on an LLNL CZ machine with `py-ujson@4.0.2` and `py-zipp@3.6.0`.

File: var/spack/repos/builtin/packages/py-sina/package.py (hunk 1 above)
Comment (doutriaux1) [reply var/spack/repos/builtin/packages/py-sina/package.py]: I built it yesterday on LC RZ

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/py-sina/package.py]: Then there are likely missing dependencies or dependency constraints (e.g., version of Python and or the dependencies perhaps).

**Comment by doutriaux1** [comment]
ok let me try again on the cz then

File: var/spack/repos/builtin/packages/py-sina/package.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        'doutriaux1',
+    ]
+
+    version('1.11.0', tag="v1.11.0")
+    version('1.10.0', tag="v1.10.0")
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-ujson', type=('build', 'run'))
+    depends_on("py-sqlalchemy", type="run")
+    depends_on("py-six", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sina/package.py]:
```suggestion
    depends_on("py-sqlalchemy", type=("build", "run"))
    depends_on("py-six", type=("build", "run"))
```
You're also missing the following deps:
```python
        'enum34;python_version<"3.4"',                                                   
        'orjson;python_version>="3.6" and platform_machine!="ppc64le"',                  
        'ujson;python_version>="3.6" and platform_machine=="ppc64le"',                   
        'ujson<4;python_version<"3.6" and platform_machine!="ppc64le"',         
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-sina/package.py (hunk 4 above)
Comment (doutriaux1) [reply var/spack/repos/builtin/packages/py-sina/package.py]: how would I specify enum34;python_version<"3.4 in spack?
orjson does not exists in spack so we have to use ujson everywhere.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-sina/package.py]: > how would I specify enum34;python_version<"3.4 in spack?

```python
depends_on('py-enum34', when='^python@:3.3', type=('build', 'run'))
```

> orjson does not exists in spack so we have to use ujson everywhere.

I'm not sure if that will work unless you're on ppc64le. It shouldn't be too hard to add a `orjson` package.

Comment (doutriaux1) [reply var/spack/repos/builtin/packages/py-sina/package.py]: I'm positive orjson is only a "nice to have" json package (it's a bit faster). orjson uses not so common build tools "rust" and "maturin". If it ever makes it into spack I'm willing to update the recipe to use it.

Comment (doutriaux1) [reply var/spack/repos/builtin/packages/py-sina/package.py]: building on CZ i now get this:
==> Bootstrapping clingo from sources
==> Error: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %gcc target=x86_64" due to the following failures:
    'github-actions' raised FetchCacheError: Multiple errors during fetching:
        Error 1: RuntimeError: Unable to read index hash https://mirror.spack.io/bootstrap/github-actions/v0.1/build_cache/index.json.hash due to SpackWebError: Download failed: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1091)>
        Error 2: RuntimeError: Unable to read index https://mirror.spack.io/bootstrap/github-actions/v0.1/build_cache/index.json due to SpackWebError: Download failed: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1091)>
    'spack-install' raised ValueError: Invalid config scope: 'bootstrap'.  Must be one of odict_keys(['_builtin', 'defaults', 'defaults/linux', 'bootstrap/linux', 'disable_modules', 'overrides-0'])
    Please run `spack -d spec zlib` for more verbose error messages

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-sina/package.py]: @alalazo can you help debug this bootstrap error?

Comment (doutriaux1) [reply var/spack/repos/builtin/packages/py-sina/package.py]: @alalazo I'm still getting this error today.

Comment (doutriaux1) [reply var/spack/repos/builtin/packages/py-sina/package.py]: @alalazo moving from quartz to surface, it now works fine. Not sure what's up with quartz....

**Comment by doutriaux1** [comment]
@tldahlgren I think we're good to go now. I looked a  bit more into orjson, it needs rust (in spack) but also maturin (not in spack) which in turns need cargo (not in spack)... So considering the very small speed advantage of orjson over ujson, I think we'll punt on orjson for now.

**Comment by adamjstewart** [comment]
My concern with orjson is that since the `setup.py` contains:
```
'orjson;python_version>="3.6" and platform_machine!="ppc64le"',                  
'ujson;python_version>="3.6" and platform_machine=="ppc64le"',                   
'ujson<4;python_version<"3.6" and platform_machine!="ppc64le"',  
```
then `orjson` is "required" when using Python 3.6 and not on pcc64le. It isn't optional, and if it isn't found, setuptools/pip will try to install it for you, downloading it externally. This is problematic on air-gapped networks where you can't download anything during the install.

You could patch this file to require `ujson` for all Python versions and OSes to get around this.

**Comment by doutriaux1** [comment]
@adamjstewart let me look into the patch. I like that idea.

**Comment by doutriaux1** [comment]
@adamjstewart I think we're good with the patch, looks like orjson was not installed via pip:
```bash
$ which python
/blah/spack/var/spack/environments/test-pysina/.spack-env/view/bin/python
$ python -c "import orjson"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'orjson'
$ python -c "import ujson"
$
```

File: var/spack/repos/builtin/packages/py-sina/no_orjson.patch [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
++++ b/python/setup.py
+@@ -51,9 +51,8 @@ setup(name='llnl-sina',
+         'six',
+         'sqlalchemy',
+         'enum34;python_version<"3.4"',
+-        'orjson;python_version>="3.6" and platform_machine!="ppc64le"',
+-        'ujson;python_version>="3.6" and platform_machine=="ppc64le"',
+-        'ujson<4;python_version<"3.6" and platform_machine!="ppc64le"',
++        'ujson;python_version>="3.6"',
++        'ujson<4;python_version<"3.6"',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sina/no_orjson.patch]: Let's just make it 'ujson' without the Python version check. This kind of stuff is only in there to workaround bugs in pip. Spack is smart enough to know that `py-ujson` 4+ only supports Python 3.

Comment (doutriaux1) [reply var/spack/repos/builtin/packages/py-sina/no_orjson.patch]: Done.

**Review by adamjstewart — APPROVED** [review]

**Comment by doutriaux1** [comment]
@tldahlgren is there anything else you want from me and that I missed?

**Comment by tldahlgren** [comment]
> @tldahlgren is there anything else you want from me and that I missed?

Not that I recall.

**Review by tldahlgren — APPROVED** [review]
